include/swift/AST/DiagnosticsSema.def

```diff
@@ -71,6 +76,8 @@ ERROR(could_not_find_value_dynamic_member_corrected,none,
 ERROR(could_not_find_value_dynamic_member,none,
       "value of type %0 has no dynamic member %2 using key path from root type %1",
       (Type, Type, DeclNameRef))
+ERROR(cannot_infer_contextual_keypath_type_specify_root,none,
+      "cannot infer contextual type for key path; consider adding an explicit root type", ())
```

Comment:
Thoughts on dropping "contextual" and "explicit"?

```
cannot infer type for key path; consider adding a root type
```

Reply:
I really not sure, I think is fine to change it, but I think the actual verbosity makes a bit more clear.
But can change no problem :)

Reply:
That's a good reason to have a discussion. I'm not trying to force people into following what I say. Here's my thought process. Please feel free to disagree and expand on why you think those terms make the diagnostic clearer. That's how we make things better. 🙂 

With "explicit root types", there's the (probably small) additional risk of an wrong turn where someone might think "what is the difference from a root type and an explicit root type" (even though what we actually meant was "adding a root type explicitly"). (FWIW, I Google-searched for "Swift key path root type" and "Swift key path explicit root type" and the former gave better search results.)

With "contextual type" it's likely that the programmer has seen an error mentioning contextual type before and the way they typically fix it is by actually adding a type annotation for a binding (or in some cases using `as`)

```
let x = .mycase -> let x : MyType = .mycase
```

So they might think "how should I add a type annotation with `:`" or they might think "should I add a ", if they hung up on the first part of the diagnostic (maybe because they don't fully understand what the second part of the diagnostic is trying to say). But both of those solutions are suboptimal.

Again, this is probably a small risk, but it's a potential thing that can go wrong.

What we really want is the person to understand that (a) there is a type inference failure (as opposed to many other kinds of failures) (b) adding a root type to the key path expressions will fix the problem. I don't think the terms "contextual" or "explicit" are justifying the potential cognitive cost, because they're not advancing any of these sub-goals IMO. The verbiage without them

```
cannot infer type for key path; consider adding a root type
```

is better IMO in that both sub-goals are tackled with less room for following an incorrect line of reasoning.

Reply:
Humm that makes sense, thank you for the detailed description :) 
I just changed, adding an extra `from context` to make it clear that it is failing to infer from the current context the expression is in. `cannot infer type for key path;` may give room to think that the compiler cannot infer key path types at all(although is obviously not true). 
The message is `cannot infer key path type from context; consider adding a root type` let me know what you think :)

Reply:
Thanks. I see your point about potentially making someone think that the compiler can't infer those types generally, that's fair. I think if someone's using a key path, it is likely that they've probably already used it in contexts where the type was inferred for them, so maybe it's not such a big risk.

Reply:
If we want to be precise, one isn't "adding" a root type (there always has to be one), only specifying that type explicitly. To avoid confusion you could write "consider explicitly adding a root type."

Reply:
I like it that way, feels precise :)